cargo: Introduce feature based compiling

cathay4t opened the pull request:
Introducing cargo features(by default, both are enabled)
 * `query_apply`
 * `gen_conf`

The code has been split to small parts to pass rust clippy check.

The general rule of this split is:
   Try our best to use use `#[cfg(feature = "query_apply")]` only in
    `lib.rs` or `mod.rs`.

To only compile the `genconf` part:

    cargo build --no-default-features \
        --features gen_conf

Github Action expanded to run cargo clippy for genconf only feature also.
Also enabled MacOS build test for genconf feature only.

qinqon commented:
Can we split the "features" code at different mods so we don't have "spaghetti" cfg ?

cathay4t committed:
tox: Fix error on missing pylint 2.15

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t commented:
@qinqon I have changed to code layout for fit this conditional compiling with `#[cfg(feature = "query_apply")]` in `lib.rs` or `mod.rs`

Please review the new patch again.

cathay4t commented:
The `dep:` feature require rust 1.60+  while RHEL 8.6 and RHEL 9.0 is still rust 1.58.

Let me change the rpmspec to enable nightly feature in that case.

Considering this branch is targeting RHEL 8.8 and 9.2 where rust will be 1.62+, so we should be OK.

cathay4t committed:
rust: Conditional compiling by cargo feature

Introducing cargo features(by default, both are enabled)
 * `query_apply`
 * `gen_conf`

The code has been split to small parts to pass rust clippy check.

The general rule of this split is:
   Try our best to use use `#[cfg(feature = "query_apply")]` only in
    `lib.rs` or `mod.rs`.

To only compile the `gen_conf` part:

    cargo build --no-default-features \
        --features gen_conf

Github Action expanded to run cargo clippy for gen_conf only feature also.
Also enabled MacOS build test for gen_conf feature only.

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t commented:
Toni tested in MacOS VM, the `gc` mode works well.

ffmancera commented on .packit.yml:
```diff
@@ -20,5 +20,3 @@ jobs:
       targets:
         - centos-stream-8-x86_64
         - centos-stream-9-x86_64
-        - epel-8-x86_64
-        - epel-9-x86_64
```
Why is this needed? Could you explain it in the commit message? Thanks!

cathay4t replied:
Commit message updated to explain why we drop those build target.

cathay4t committed:
packit: Remove the PR copr build for EPEL8 and EPEL9

Currently, EPEL8 is based on RHEL 8.6 and EPEL9 is based on RHEL 9.0
both are shipping rust 1.58 which does not support
`namespaced-features`. That is required for follow up conditional
compiling patch.

Will enable it once EPEL switch to RHEL 8.7 and 9.1 where rust is 1.62.

Signed-off-by: Gris Ge <fge@redhat.com>
